File: scripts/cleanup-data.js
const postgres = require("postgres");
const sql = postgres(process.env.DATABASE_URL, { max: 1 });

async function main() {
  // Find and remove dummy/placeholder players that create false connections
  const dummies = await sql`
    SELECT id, name FROM players
    WHERE name IN ('NN', 'N.N.', 'N.N', '?', '??', 'Unknown', 'unknown', 'Comp', 'Computer')
    OR name LIKE 'Comp %'
    OR name LIKE 'Team %'
  `;

  console.log(`Found ${dummies.length} dummy players to remove:`);
  for (const d of dummies) {
    console.log(`  ${d.id}: "${d.name}"`);
  }

  if (dummies.length > 0) {
    const ids = dummies.map(d => d.id);
    // Delete opponent pairs involving dummy players
    for (const id of ids) {
      const del = await sql`DELETE FROM opponents WHERE player_a_id = ${id} OR player_b_id = ${id}`;
      console.log(`  Deleted pairs for player ${id}`);
    }
    // Delete dummy players
    for (const id of ids) {
      await sql`DELETE FROM players WHERE id = ${id}`;
    }
    console.log(`Removed ${ids.length} dummy players and their connections`);
  }

  // Check remaining stats
  const stats = await sql`SELECT (SELECT count(*) FROM players) as p, (SELECT count(*) FROM opponents) as o`;
  console.log(`\nRemaining: ${stats[0].p} players, ${stats[0].o} opponent pairs`);
}

main().catch(console.error);

File: etl/cleanup_data.py
from __future__ import annotations
"""Remove dummy/placeholder players that create false connections.

Deletes players named 'NN', 'Computer', 'Comp *', 'Team *', etc., along
with all their opponent pairs.

Usage:
    DATABASE_URL=... python etl/cleanup_data.py
"""

import sys
from pathlib import Path

import psycopg.rows

sys.path.insert(0, str(Path(__file__).parent.parent))
from etl.db import get_conn


def main() -> None:
    conn = get_conn()
    with conn.cursor(row_factory=psycopg.rows.dict_row) as cur:
        cur.execute("""
            SELECT id, name FROM players
            WHERE name IN ('NN', 'N.N.', 'N.N', '?', '??', 'Unknown', 'unknown', 'Comp', 'Computer',
                           'Stockfish', 'Komodo', 'Houdini', 'Fritz', 'Deep Blue', 'AlphaZero',
                           'Leela', 'LC0', 'Lc0', 'Crafty', 'Rybka', 'Shredder', 'Junior',
                           'Deep Junior', 'Deep Fritz', 'Hydra', 'Toga', 'Zappa')
               OR name LIKE 'Comp %%'
               OR name LIKE 'Team %%'
               OR name LIKE 'Engine %%'
               OR name LIKE 'Stockfish %%'
               OR name LIKE 'Komodo %%'
               OR name LIKE 'Houdini %%'
               OR name LIKE 'Fritz %%'
               OR name LIKE 'Deep %%'
               OR name LIKE 'Leela %%'
               OR name LIKE 'LC0 %%'
               OR name LIKE 'Lc0 %%'
               OR name LIKE 'Rybka %%'
               OR name LIKE 'Shredder %%'
        """)
        dummies = cur.fetchall()

    print(f"Found {len(dummies)} dummy players to remove:")
    for d in dummies:
        print(f"  {d['id']}: \"{d['name']}\"")

    if dummies:
        with conn.cursor() as cur:
            for d in dummies:
                cur.execute("DELETE FROM opponents WHERE player_a_id = %s OR player_b_id = %s", (d['id'], d['id']))
                cur.execute("DELETE FROM players WHERE id = %s", (d['id'],))
                print(f"  Deleted pairs and player {d['id']}")
        conn.commit()
        print(f"Removed {len(dummies)} dummy players and their connections")

    with conn.cursor(row_factory=psycopg.rows.dict_row) as cur:
        cur.execute("SELECT (SELECT count(*) FROM players) AS p, (SELECT count(*) FROM opponents) AS o")
        stats = cur.fetchone()
    print(f"\nRemaining: {stats['p']:,} players, {stats['o']:,} opponent pairs")
    conn.close()


if __name__ == "__main__":
    main()
